libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Util/SFSDKOAuth2.m

```diff
@@ -0,0 +1,416 @@
+    if (self = [super init]) {
+        _error = error;
+        
+    }
+    return self;
+}
+
+@end
+
+@implementation SFSDKOAuthTokenEndpointResponse
+
+- (instancetype)initWithError:(NSError *)error {
+    if (self = [super init]) {
+        _error = [[SFSDKOAuthTokenEndpointErrorResponse alloc] initWithError:error];
+        
+    }
+    return self;
+}
+
+- (instancetype)initWithDictionary:(NSDictionary *)nvPairs parseAdditionalFields:(NSArray<NSString *> *)additionalOAuthParameterKeys {
+    
+    if (self = [super init]) {
+        _values = nvPairs;
+        _additionalOAuthParameterKeys = additionalOAuthParameterKeys;
+        
+        if (additionalOAuthParameterKeys) {
+            NSMutableDictionary * parsedValues = [NSMutableDictionary dictionaryWithCapacity:_additionalOAuthParameterKeys.count];
+            for(NSString * key in self.additionalOAuthParameterKeys) {
+                id obj = nvPairs[key];
+                if (obj) {
+                    parsedValues[key] = obj;
+                }
+            }
+            _additionalOAuthFields = parsedValues;
+        }
+       
+        if (nvPairs[kSFOAuthScope]) {
+            NSString *rawScope = nvPairs[kSFOAuthScope];
+            _scopes = [rawScope componentsSeparatedByString:@" "];
+        }
+        
+        if (nvPairs[kSFOAuthError]) {
+            _error = [[SFSDKOAuthTokenEndpointErrorResponse alloc] initWithError:nvPairs[kSFOAuthError] description:nvPairs[kSFOAuthErrorDescription]];
+        }
+        
+    }
+    return self;
+}
+
+- (BOOL)hasError {
+    return (self.error != nil);
+}
+
+- (NSString *)accessToken {
+    return self.values[kSFOAuthAccessToken];
+}
+
+- (NSString *)refreshToken {
+    return self.values[kSFOAuthRefreshToken];
+}
+
+- (NSDate *)issuedAt {
+     return [SFSDKOAuth2 timestampStringToDate:self.values[kSFOAuthIssuedAt]];
+}
+
+- (NSURL *)instanceUrl {
+    if (self.values[kSFOAuthInstanceUrl]) {
+        return [NSURL URLWithString:self.values[kSFOAuthInstanceUrl]];
+    }
+    return nil;
+}
+
+- (NSURL *)identityUrl {
+    if (self.values[kSFOAuthId]) {
+        return [NSURL URLWithString:self.values[kSFOAuthId]];
+    }
+    return nil;
+}
+
+- (NSString *)communityId {
+     return self.values[kSFOAuthCommunityId];
+}
+
+- (NSString *)signature {
+    return self.values[@"signature"];
+}
+
+- (NSURL *)communityUrl {
+    if (_values[kSFOAuthCommunityUrl]) {
+        return [NSURL URLWithString:self.values[kSFOAuthCommunityUrl]];
+    }
+    return nil;
+}
+
+- (NSDictionary *)asDictionary {
+    return self.values;
+}
+@end
+
+@implementation SFSDKOAuth2
+
+- (void)accessTokenForApprovalCode:(SFSDKOAuthTokenEndpointRequest *)endpointReq completion:(void (^)(SFSDKOAuthTokenEndpointResponse *)) completionBlock {
+    
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:endpointReq.serverURL cachePolicy:NSURLRequestReloadIgnoringCacheData timeoutInterval:endpointReq.timeout];
```

Comment:
Should we use `SFNetwork` here? Or does it have other implications such as auto-refresh?